### ISSUE TITLE
[java] MissingStaticMethodInNonInstantiatableClass: False negative when method in nested class returns null

### ISSUE THREAD

**Issue by adangel** [pmd/pmd#6343]
**Affects PMD Version:** 7.19.0

**Rule:** [MissingStaticMethodInNonInstantiatableClass](https://docs.pmd-code.org/latest/pmd_rules_java_errorprone.html#missingstaticmethodinnoninstantiatableclass)

**Description:**
When there is a nested class that has a method that returns null, then the rule considers this wrongly as a builder method which returns an instance of the outer class.

This is due to currently that `null` is considered to be compatible with any type, which is not correct. With the changes in #6292 this will be corrected.

**Code Sample demonstrating the issue:**

```java
public class MyClass {

    private MyClass() {
    }

    public static class NotABuilder {

        private NotABuilder() {}

        public NotABuilder notABuilder() {
            return null;
        }
    }
}
```

**Expected outcome:**

PMD should report a violation at line 1, but doesn't. This is a false-negative.
